capsules/src/lpm013m126.rs

```diff
@@ -0,0 +1,487 @@
+                        );
+                        let l = FrameBuffer::with_raw_rows(frame_buffer, 0, 1);
+                        let res = self.spi.read_write_bytes(
+                            l, //FrameBuffer::with_raw_rows(frame_buffer, 0, 1),
+                            None, 2,
+                        );
+
+                        let (res, new_state) = match res {
+                            Ok(()) => (Ok(()), State::InitializingPixelMemory),
+                            Err((e, buf, _)) => {
+                                self.frame_buffer.replace(FrameBuffer::new(buf));
+                                (Err(e), State::Bug)
+                            }
+                        };
+                        self.state.set(new_state);
+                        res
+                    }
+                }
+            }
+            _ => Err(ErrorCode::ALREADY),
+        }
+    }
+
+    fn arm_alarm(&self) {
+        // Datasheet says 120Hz or more often flipping is required
+        // for transmissive mode.
+        let delay = self.alarm.ticks_from_ms(500);
+        self.alarm.set_alarm(self.alarm.now(), delay);
+    }
+}
+
+impl<'a, A: Alarm<'a>, P: Pin, S: SpiMasterDevice> Screen for Lpm013m126<'a, A, P, S> {
+    fn get_resolution(&self) -> (usize, usize) {
+        (176, 176)
+    }
+
+    fn get_pixel_format(&self) -> ScreenPixelFormat {
+        ScreenPixelFormat::Mono
+    }
+
+    fn get_rotation(&self) -> ScreenRotation {
+        ScreenRotation::Normal
+    }
+
+    fn set_write_frame(
+        &self,
+        x: usize,
+        y: usize,
+        width: usize,
+        height: usize,
+    ) -> Result<(), ErrorCode> {
+        let rows = 176;
+        let columns = 176;
+        if y >= rows || y + height >= rows || x >= columns || x + width >= columns {
+            return Err(ErrorCode::INVAL);
+        }
+
+        let frame = WritePosition {
+            row: y as u16,
+            column: x as u16,
+            width: width as u16,
+            height: height as u16,
+        };
+
+        let mut new_state = None;
+        let ret = match self.state.get() {
+            State::Uninitialized => Err(ErrorCode::OFF),
+            State::InitializingPixelMemory | State::InitializingRest => Err(ErrorCode::BUSY),
+            State::Idle(..) => {
+                new_state = Some(State::Idle(frame));
+                Ok(())
+            }
+            State::Writing(..) => {
+                new_state = Some(State::Writing(frame));
+                Ok(())
+            }
+            State::Bug => Err(ErrorCode::FAIL),
+        };
+
+        if let Some(new_state) = new_state {
+            self.state.set(new_state);
+        }
+        // Thankfully, this is the only command that results in the callback,
+        // so there's no danger that this will get attributed
+        // to a command that's not finished yet.
+        self.client.map(|client| client.command_complete(Ok(())));
+        ret
```

Comment:
This doesn't quite look right. Callbacks cannot be called from HIL commands, as this breaks Tock's control flow model (i.e. the caller of `screen.set_write_frame()` can expect that its code completes before the callback function is called). If this callback is needed then it needs to be triggered originating from another callback (such as alarm.fired() for example).

Reply:
Is there some "idle" callback triggered when the kernel has nothing else to schedule? Should I try adding one?

Getting a timer involved for what should be synchronous sounds like an overkill, both in terms of code as well as resources.

Reply:
You can use a deferred call, similar to the one used by this driver (https://github.com/WyliodrinEmbeddedIoT/getting-started-with-secure-embedded-systems/blob/main/chapter_10/kernel/drivers/src/led_matrix_text.rs), to schedule an upcall after your function return.